### ISSUE TITLE
Hexagon Delegate crashes on max pooling operation

### ISSUE THREAD

**Issue by yakovdan** [tensorflow/tensorflow#51804]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Snapdragon 855 dev platform
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.6.0
- Python version: 3.8.0
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
After converting YOLOv4 from Tensorflow to TfLite, the model runs well on mobile device, on CPU.
When running on DSP using hexagon delegate, I get the following error:

INFO: Initialized TensorFlow Lite runtime.
INFO: TfLiteHexagonDelegate delegate: 535 nodes delegated out of 568 nodes with 2 partitions.
INFO: Replacing 535 node(s) with delegate (TfLiteHexagonDelegate) node, yielding 5 partitions.

----------------
Timestamp: Sun May 30 16:21:13 2021

Log
hexagon/ops/src/op_maxpool_d32.c:567:insufficient width padding
hexagon/src/execute.c:167:execute() failed on node id=689 err=-1
hexagon/src/interface.c:1297:fail in execute_inner()

----------------
ERROR: Failed: Failed to execute graph..
ERROR: Node number 568 (TfLiteHexagonDelegate) failed to invoke.

**Describe the expected behavior**

I expect it to run successfully. 

**Standalone code to reproduce the issue**
See attached model. 

**Other info / logs** 
I've narrowed it down to a single max pool operation that is later concatenated. The operation in question takes a 1x13x13x512 input tensor and computes tf.nn.max_pool(input_data, ksize=13,padding = "SAME", strides=1). other instances of this operator with ksize = 9 and ksize =5 (all other parameters are the same) work just fine. 

Another thing I noticed is that depending on what else is going on in the model, sometimes the converter emits a TfLite model that works. However, it is intermittent and I was unable to reproduce this behavior reliably 

See model attached [here](https://drive.google.com/file/d/1sAh0i-hV2936tVGKliyX5sMcWbfxqU3-/view?usp=sharing).

**Comment by yakovdan** [comment]
In case you want replicate creating the quantized model:

Here's a [zip file](https://drive.google.com/file/d/1J3kg2Ul0w0Dw9w29nU4MFRKD7tjv47B8/view?usp=sharing) with the TensorFlow saved_model (./checkpoints/yolov4-416), the original model definitions (./core), the representative dataset based on COCO (./data/dataset/val2017) and the conversion script: convert_tflite.py 
which you run like this:
python convert_tflite.py --weights ./checkpoints/yolov4-416 --output ./checkpoints/yolov4-416-int8.tflite --quantize_mode int8 --dataset ./data/dataset/val2017.txt

you can find the problematic max_pool being used at ./core/backbone.py, line 99.

**Comment by srjoglekar246** [comment]
> Another thing I noticed is that depending on what else is going on in the model, sometimes the converter emits a TfLite model that works. However, it is intermittent and I was unable to reproduce this behavior reliably

This is certainly weird. What do you mean by 'intermittent'? Does it:

1. Not convert a model sometimes, with a conversion error? OR
2. Converts a model that runs on Hexagon just fine?

**Comment by yakovdan** [comment]
I meant the second option. Sometimes I get a model that runs fine, sometimes I get a model that crashes with the error reported above. One caveat - I haven't yet verified the correctness of the output when executed on Hexagon, just that execution terminates without runtime errors (when it runs fine). However, the same model runs fine on PC, always - and the output is correct as well.

**Comment by yakovdan** [comment]
Hi, are there any updates on this? Is there anything I can do to help debug this?

**Comment by srjoglekar246** [comment]
Sorry for the delay, we had a long weekend in the US. If possible can you paste the models that 1) runs fine on Hexagon & 2) gives an error with Hexagon? If you only intermittently see Hexagon errors, then there might be some issue with the converter. I can compare the two models & see what is different between the two.

**Comment by yakovdan** [comment]
I've tried many times to recreate a model that runs on Hexagon, but I couldn't. They all failed to run as I've described above. It's likely the intermittency I've seen is just confusion on my part. 

Can we discuss why max_pool with size (13,13) fails?

In the meantime, I've replaced the max_pool operator with kernel size (13,13) with a chain of two smaller max_pool operators that work fine - (9,9) and (5,5), and a multiplication in the following manner:
```
### replaced this:
  #input_data = tf.concat([tf.nn.max_pool(input_data, ksize=13, padding='SAME', strides=1), tf.nn.max_pool(input_data, ksize=9, 
  #padding='SAME', strides=1)
   # , tf.nn.max_pool(input_data, ksize=5, padding='SAME', strides=1), input_data], axis=-1)

### with this:
    step1 = tf.nn.max_pool(input_data, ksize=9, padding='VALID', strides=1)
    step2 = tf.nn.max_pool(step1, ksize=5, padding='VALID', strides=1)
    step3=tf.math.multiply(tf.constant(1.0, shape=(1, 13, 13, 512), name='test'), step2)
    
    input_data = tf.concat([step3, tf.nn.max_pool(input_data, ksize=9, padding='SAME', strides=1)
                            , tf.nn.max_pool(input_data, ksize=5, padding='SAME', strides=1), input_data], axis=-1)
```

This works. But is there a way to resolve this without ugly workarounds?

**Comment by srjoglekar246** [comment]
The error comes from within the Hexagon kernel ([see code here](https://github.com/XiaoMi/nnlib/blob/master/hexagon/ops/src/op_maxpool_d32.c#L566)), so you can probably follow the logic there (with your op's IO dimensions & arguments) to get a better idea of why this particular instance of the op is failing. I am a bit swamped at the moment, so might not have the time to dig myself. Could you help take a look?


I am glad you atleast found a workaround for your use-case :-)

**Comment by srjoglekar246** [comment]
For now, we have disabled accepting MaxPool operations if padding is SAME and filter size > 12 (which is where we see errors in our unit tests too). It is unfortunately not very likely that this will be fixed in the Hexagon SDK code very soon, but the TFLite delegate interface should work fine & run the rest of the graph correctly even if such ops are present.

**Comment by karimnosseir** [comment]
Closing the issue, as this shouldn't be happening now.
